File: v2/python/anhurdb/storage/filesystem.py
"""
Compressed payload storage for AnhurDB records.

Handles reading gzip-compressed JSON files from the NFS-backed storage
layer, with a REST API fallback when direct disk access is unavailable.

Security:
    - Path traversal protection: ``tenant_id``, ``uuid``, and ``record_id``
      are validated against directory escape characters (``..``, ``/``, ``\\``)
      before constructing file paths.
    - Symlink resolution: ``os.path.realpath`` is used to verify the resolved
      path stays within the configured ``base_path``.
"""

import os
import gzip
import json
import re
from typing import Any, Dict, Optional

import requests

# Regex that rejects any path component containing directory traversal
# or absolute path characters. Only allows alphanumeric, hyphens, underscores,
# and dots (not leading dots).
_SAFE_PATH_COMPONENT = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9._-]*$")


def _validate_path_component(value: str, name: str) -> None:
    """
    Validate that a path component is safe against directory traversal.

    Rejects:
        - Empty strings
        - Components containing ``..``
        - Components containing ``/`` or ``\\``
        - Components starting with ``.``
        - Components with null bytes

    Args:
        value: The path component to validate.
        name:  Human-readable name for error messages (e.g. "tenant_id").

    Raises:
        ValueError: If the component is unsafe.
    """
    if not value:
        raise ValueError(f"{name} cannot be empty")
    if "\x00" in value:
        raise ValueError(f"{name} contains null byte")
    if ".." in value:
        raise ValueError(f"{name} contains directory traversal sequence '..'")
    if "/" in value or "\\" in value:
        raise ValueError(f"{name} contains path separator")
    if not _SAFE_PATH_COMPONENT.match(value):
        raise ValueError(
            f"{name} contains invalid characters: only alphanumeric, "
            f"hyphens, underscores, and non-leading dots are allowed"
        )


class FileStorage:
    """
    Read compressed cognitive payload files from the NFS storage layer.

    AnhurDB stores record content as gzip-compressed JSON files at
    ``{base_path}/{tenant_id}/{uuid}/{record_id}.gz``.

    Args:
        base_path: Root directory for record storage (e.g. ``/data/storage``).
    """

    def __init__(self, base_path: str):
        self.base_path = os.path.realpath(base_path)

    def build_path(self, tenant_id: str, uuid: str, record_id: int) -> str:
        """
        Construct and validate the filesystem path for a record payload.

        All components are validated against path traversal before joining.
        The resulting path is verified to stay within ``base_path`` via
        ``os.path.realpath`` comparison.

        Args:
            tenant_id: Hex-encoded tenant identifier.
            uuid:      Session UUID.
            record_id: Numeric record ID.

        Returns:
            Absolute path to the ``.gz`` file.

        Raises:
            ValueError: If any component contains traversal characters.
        """
        _validate_path_component(tenant_id, "tenant_id")
        _validate_path_component(uuid, "uuid")

        if record_id < 0:
            raise ValueError("record_id must be non-negative")

        path = os.path.join(self.base_path, tenant_id, uuid, f"{record_id}.gz")

        # Final defense: verify resolved path is under base_path.
        real_path = os.path.realpath(path)
        if not real_path.startswith(self.base_path + os.sep):
            raise ValueError(
                f"Resolved path {real_path} escapes base directory {self.base_path}"
            )

        return real_path

    def read_json(self, tenant_id: str, uuid: str, record_id: int) -> Dict[str, Any]:
        """
        Read a gzip-compressed JSON payload from disk.

        Args:
            tenant_id: Hex-encoded tenant identifier.
            uuid:      Session UUID.
            record_id: Numeric record ID.

        Returns:
            Parsed JSON dict.

        Raises:
            FileNotFoundError: If the file does not exist.
            ValueError: If path components are unsafe.
        """
        path = self.build_path(tenant_id, uuid, record_id)
        if not os.path.exists(path):
            raise FileNotFoundError(f"Record payload not found: {path}")

        with gzip.open(path, "rt", encoding="utf-8") as f:
            return json.load(f)

    def read_json_with_fallback(
        self,
        tenant_id: str,
        uuid: str,
        record_id: int,
        api_url: Optional[str] = None,
        api_key: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Read payload from disk, falling back to the REST API if unavailable.

        Tries direct disk read first (fastest). If the file doesn't exist
        (e.g. due to NFS isolation), transparently falls back to the
        ``/api/v1/records/{id}/content`` REST endpoint.

        Args:
            tenant_id: Hex-encoded tenant identifier.
            uuid:      Session UUID.
            record_id: Numeric record ID.
            api_url:   AnhurDB server URL for fallback (optional).
            api_key:   API key for fallback authentication (optional).

        Returns:
            Parsed JSON dict.

        Raises:
            FileNotFoundError: If both disk and API fallback fail.
            ValueError: If path components are unsafe.
        """
        path = self.build_path(tenant_id, uuid, record_id)
        if os.path.exists(path):
            with gzip.open(path, "rt", encoding="utf-8") as f:
                return json.load(f)

        # Fallback to REST API.
        if not api_url:
            raise FileNotFoundError(
                f"Record payload not found at {path} and no API fallback URL."
            )

        headers: Dict[str, str] = {}
        if api_key:
            headers["X-API-Key"] = api_key
        if tenant_id:
            headers["X-Tenant-ID"] = tenant_id

        # Validate record_id is numeric to prevent injection in URL.
        url = f"{api_url.rstrip('/')}/api/v1/records/{int(record_id)}/content"
        resp = requests.get(url, headers=headers, timeout=30)
        if not resp.ok:
            raise FileNotFoundError(
                f"Record not found locally and API fallback returned "
                f"HTTP {resp.status_code}"
            )

        return resp.json()

File: v2/golang/storage/filesystem.go
package storage

import (
	"compress/gzip"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// fallbackHTTPTimeout bounds the REST fallback request.
//
// Junior Tip [security/availability]: an http.Client with no Timeout will wait
// FOREVER on a stalled connection (no read/write deadline). A hung fallback
// silently blocks the caller's goroutine, which in the pipeline cascades into
// stuck workers and eventual data-loss-by-timeout upstream. 30s matches the
// Python (timeout=30) and TypeScript (AbortSignal.timeout(30_000)) SDKs so all
// three behave identically.
const fallbackHTTPTimeout = 30 * time.Second

// validatePathComponent rejects a path component that could escape the storage
// root via directory traversal.
//
// Junior Tip [security]: tenantID and uuid flow into a filesystem path. If a
// caller (or a compromised upstream) passes "../../etc" the join would resolve
// OUTSIDE BasePath and read/leak arbitrary files. We reject "..", path
// separators ("/" and "\\"), and null bytes BEFORE building any path. This
// mirrors the Python SDK's _validate_path_component so the 3 SDKs enforce the
// same contract. We fail LOUD (return an error) rather than silently sanitising
// — a traversal attempt is a bug or an attack, never normal input.
func validatePathComponent(value, name string) error {
	if value == "" {
		return fmt.Errorf("%s cannot be empty", name)
	}
	if strings.Contains(value, "\x00") {
		return fmt.Errorf("%s contains null byte", name)
	}
	if strings.Contains(value, "..") {
		return fmt.Errorf("%s contains directory traversal sequence '..'", name)
	}
	if strings.Contains(value, "/") || strings.Contains(value, "\\") {
		return fmt.Errorf("%s contains path separator", name)
	}
	return nil
}

// FileStorage handles reading and writing compressed cognitive payload files.
type FileStorage struct {
	BasePath string
}

// NewFileStorage creates a new FileStorage reference.
func NewFileStorage(basePath string) *FileStorage {
	return &FileStorage{
		BasePath: basePath,
	}
}

// BuildPath constructs the expected file path for a tenant record.
//
// Junior Tip [security]: tenantID and uuid are validated for path traversal
// BEFORE the join (see validatePathComponent). Returning an error rather than a
// bare string is deliberate — callers MUST surface a traversal attempt instead
// of letting a malicious path silently resolve outside BasePath. Matches the
// Python SDK's build_path, which raises ValueError on the same conditions.
func (f *FileStorage) BuildPath(tenantID, uuid string, recordID int) (string, error) {
	if validateErr := validatePathComponent(tenantID, "tenantID"); validateErr != nil {
		return "", validateErr
	}
	if validateErr := validatePathComponent(uuid, "uuid"); validateErr != nil {
		return "", validateErr
	}
	if recordID < 0 {
		return "", fmt.Errorf("recordID must be non-negative, got %d", recordID)
	}
	return filepath.Join(f.BasePath, tenantID, uuid, fmt.Sprintf("%d.gz", recordID)), nil
}

// Read reads and decompresses a gzip file from the local filesystem.
// Returns the decompressed bytes.
func (f *FileStorage) Read(tenantID, uuid string, recordID int) ([]byte, error) {
	path, pathErr := f.BuildPath(tenantID, uuid, recordID)
	if pathErr != nil {
		return nil, pathErr
	}
	return readGzipFile(path)
}

// readGzipFile opens a .gz file and returns decompressed content.
func readGzipFile(path string) ([]byte, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	gr, err := gzip.NewReader(file)
	if err != nil {
		return nil, fmt.Errorf("reading gzip header %s: %w", path, err)
	}
	defer gr.Close()

	return io.ReadAll(gr)
}

// ReadWithFallback first attempts to read and decompress from disk.
// If the path is inaccessible (e.g., due to strict V8 Hex isolation),
// it falls back to the REST API.
func (f *FileStorage) ReadWithFallback(tenantID, uuid string, recordID int, apiURL, apiKey string) ([]byte, error) {
	path, pathErr := f.BuildPath(tenantID, uuid, recordID)
	if pathErr != nil {
		return nil, pathErr
	}
	data, readErr := readGzipFile(path)
	if readErr == nil {
		// Found and decompressed locally
		return data, nil
	}

	// Local file not found or corrupted, fallback to REST over HTTP
	if apiURL == "" {
		return nil, fmt.Errorf("record not found locally at %s and no API URL provided for fallback", path)
	}

	url := fmt.Sprintf("%s/api/v1/records/%d/content", strings.TrimRight(apiURL, "/"), recordID)
	req, reqErr := http.NewRequest("GET", url, nil)
	if reqErr != nil {
		return nil, reqErr
	}

	if apiKey != "" {
		req.Header.Set("X-API-Key", apiKey)
	}
	if tenantID != "" {
		req.Header.Set("X-Tenant-ID", tenantID)
	}

	// Junior Tip [security]: two hardening properties on this client.
	//  1. Timeout caps the whole request so a stalled server can never hang the
	//     caller forever (the default &http.Client{} has NO timeout).
	//  2. CheckRedirect refuses to follow 3xx responses. The default client
	//     follows redirects AND re-sends our headers — including the X-API-Key
	//     credential — to whatever Location the server names, which could be a
	//     different (attacker-controlled) origin. http.ErrUseLastResponse makes
	//     Do() return the 3xx response as-is instead of following it, so the
	//     credential never crosses an origin boundary. Matches the TypeScript
	//     SDK's `redirect: "error"` and the Python SDK's allow_redirects=False.
	client := &http.Client{
		Timeout: fallbackHTTPTimeout,
		CheckRedirect: func(redirectReq *http.Request, viaRequests []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}
	resp, doErr := client.Do(req)
	if doErr != nil {
		return nil, doErr
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API fallback failed with status %d", resp.StatusCode)
	}

	return io.ReadAll(resp.Body)
}
